lstchain/reco/utils.py

```diff
@@ -778,6 +781,127 @@ def get_geomagnetic_delta(zen, az, geomag_dec=None, geomag_inc=None, time=None):
+    determine from it the intensity at which the 50% of the peak rate 
+    is reached on the rising edge.
+    
+    Parameters
+    ----------
+    data: pandas DataFrame or astropy.table.QTable of dl2 events.
+
+    Returns
+    -------
+    xmax: (float) Intensity (in p.e.) at which the peak of dR/dI is reached 
+    
+    ymax: (float) Value of dR/dI at peak (events/s/p.e.)
+
+    x50: (float) Intensity (in p.e.) at which 50% ofthe peak of dR/dI is reached
+    on the rising edge of the peak. Will be nan if not found in the search range
+
+    y50: (float) 50% of the peak dR/dI (events/s/p.e.). Will be nan if not found
+    in the search range
+
+    bincenters: (array) the centers (in log10(intensity)) of the bins of 
+    the intensity histogram used in the calculation (p.e.)
+
+    nevents: (array) the bin contents of the histogram used in the calculation
+    (events/s/p.e.)
+    
+    """
+    
+    # p.e. range where to look for the "true" cosmics peak
+    min_intensity = 25 
+    max_intensity = 1000  
+    # At lower intensity there are occasioanally fake peaks from stars, 
+    # satellites, meteors...
+    # At higher intensity there are sometimes peak from external light 
+    # sources, and from mis-tagged FF events
+    step = 0.02 # in log10(intensity), for histogramming & peak finding
+    
+    efftime, _ = get_effective_time(data)
+
+    nbins = 1 + int( (np.log10(max_intensity) - np.log10(min_intensity)) / step)
+    bins = np.logspace(np.log10(min_intensity), 
+                       np.log10(max_intensity), nbins)
+    binwidth = np.diff(bins)
+    nevents, _ = np.histogram(data['intensity'], bins=bins)
+    nevents = nevents.astype('float') / (binwidth*efftime.to_value(u.s))
+    bincenters = (bins[1:]*bins[:-1])**0.5 # geometrical mean (log bin center)
+ 
+    peaks, properties = find_peaks(np.log10(nevents), 
+                                   prominence=0.04, # ~10% in log10 scale
+                                   width=int(0.1/step)) # ~25% in log10 scale
+
+    # If no peak is found, nans are returned (except for the histogram data):
+    if len(peaks) == 0:
+        return np.nan, np.nan, np.nan, np.nan, bincenters, nevents
+    
+    xmax = bincenters[peaks.max()] # The peak at highest intensity (spurious peaks sometimes at low values)
+    ymax = nevents[peaks.max()]
+
+    # Find the intensity for which 50% of peak dR/dI is reached, moving down from the peak, in finer steps
+    # for better precision:
+    finestep = 0.001 # in log10(intensity)
+    xx = np.logspace(np.log10(min_intensity), 
+                     np.log10(xmax), 
+                     1+int((np.log10(xmax) - np.log10(min_intensity))/finestep))[::-1]
+
+    # Linear interpolation in dR/dI (differential rate) vs. log10(intensity/p.e.):    
+    yy = np.interp(np.log10(xx), np.log10(bincenters), nevents)
+
+    x50 = np.nan
+    y50 = np.nan
+    for xxx, yyy in zip(xx, yy):
```

Comment:
wouldn't it be better to use vectorize this instead of a for loop?

Reply:
This takes a completely negligible amount of time, and is run once per run